src/blockdev.rs

```diff
@@ -0,0 +1,208 @@
+            break;
+        }
+    }
+    if parents.is_empty() {
+        bail!("no parent devices found for {}", device);
+    }
+    Ok(parents)
+}
+
+#[context("get backing devices from mountpoint boot")]
+pub fn get_backing_devices<P: AsRef<Path>>(target_root: P) -> Result<Vec<String>> {
+    let target_root = target_root.as_ref();
+    let bootdir = target_root.join("boot");
+    if !bootdir.exists() {
+        bail!("{} does not exist", bootdir.display());
+    }
+    let bootdir = openat::Dir::open(&bootdir)?;
+    let fsinfo = crate::filesystem::inspect_filesystem(&bootdir, ".")?;
+    // Find the real underlying backing device for the root.
+    let backing_devices = find_parent_devices(&fsinfo.source)
+        .with_context(|| format!("while looking for backing devices of {}", fsinfo.source))?;
+    log::debug!("Find backing devices: {backing_devices:?}");
+    Ok(backing_devices)
+}
+
+#[context("Listing parttype for device {device}")]
+fn list_dev(device: &str) -> Result<BlockDevices> {
+    let mut cmd = Command::new("lsblk");
+    cmd.args([
+        "--json",
+        "--output",
+        "PATH,PTTYPE,PARTTYPE,PARTTYPENAME",
+        device,
+    ]);
+    let output = util::cmd_output(&mut cmd)?;
+    // Parse the JSON string into the `BlockDevices` struct
+    let Ok(devs) = serde_json::from_str::<BlockDevices>(&output) else {
+        bail!("Could not deserialize JSON output from lsblk");
+    };
+    Ok(devs)
+}
+
+/// Find esp partition on the same device
+pub fn get_esp_partition(device: &str) -> Result<Option<String>> {
+    let dev = list_dev(&device)?;
+    // Find the ESP part on the disk
+    for part in dev.blockdevices {
+        if part.is_esp_part() {
+            return Ok(Some(part.path));
+        }
+    }
+    log::debug!("Not found any esp partition");
+    Ok(None)
+}
+
+/// Find all ESP partitions on the backing devices with mountpoint boot
+#[allow(dead_code)]
+pub fn find_colocated_esps<P: AsRef<Path>>(target_root: P) -> Result<Vec<String>> {
+    // first, get the parent device
+    let backing_devices =
+        get_backing_devices(&target_root).with_context(|| "while looking for colocated ESPs")?;
+
+    // now, look for all ESPs on those devices
+    let mut esps = Vec::new();
+    for parent_device in backing_devices {
+        if let Some(esp) = get_esp_partition(&parent_device)? {
+            esps.push(esp)
+        }
+    }
+    log::debug!("Find esp partitions: {esps:?}");
+    Ok(esps)
+}
+
+/// Find bios_boot partition on the same device
+pub fn get_bios_boot_partition(device: &str) -> Result<Option<String>> {
+    let dev = list_dev(&device)?;
+    // Find the BIOS BOOT part on the disk
+    for part in dev.blockdevices {
+        if part.is_bios_boot_part() {
+            return Ok(Some(part.path));
+        }
+    }
+    log::debug!("Not found any bios_boot partition");
+    Ok(None)
+}
+
+/// Find all bios_boot partitions on the backing devices with mountpoint boot
+#[allow(dead_code)]
```

Comment:
I'm fine with this as is but I think it'd be slightly cleaner if this was part of the second patch